### PR TITLE
Feature/set interior vehicle data

### PR DESCRIPTION
**Pull request by AByzhynar**
- Changed method ApplicationManagerImpl::ValidateMessageBySchema
   to improve messages validation.
   Now it returns result code instead of bool
- GetInteriorVehicleData RPC changed according to new implementation of ValidateMessageBySchema
   method.
   Removed redundant function
- Implemented SetInteriorVehicleData request base functionality

### CODE REVIEW

File: src/components/application_manager/include/application_manager/service.h [hunk 1]
```diff
@@ -218,7 +226,8 @@ class Service {
    * @param message message for validation
    * @return true if message is valid according to schema, otherwise false
    */
-  virtual bool ValidateMessageBySchema(const Message& message) = 0;
+  virtual MessageValidationResult ValidateMessageBySchema(
+      const Message& message) = 0;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/service.h:230]: @AByzhynar why do you need `MessageValidationResult` return result for this function? As I could see you are using comparison with `SUCCESS` only so bool would be enough here

File: src/components/can_cooperation/src/commands/get_interior_vehicle_data_request.cc [hunk 2]
```diff
@@ -70,16 +72,17 @@ void GetInteriorVehicleDataRequest::OnEvent(
...
+  LOG4CXX_DEBUG(logger_,
+                "HMI response validation result is " << validate_result);
   const Json::Value value =
       MessageHelper::StringToValue(hmi_response.json_message());
   std::string result_code;
   std::string info;
   bool success = validate_result && ParseResultCode(value, result_code, info);
 
-  const bool is_generic_error =
-      (!validate_result) || (success && false == CheckForRequestedData(value));
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/src/commands/get_interior_vehicle_data_request.cc:81]: @AByzhynar why do this checking was removed? Is it enough to validate by schema?

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+class SetInteriorVehicleDataRequest : public BaseCommandRequest {
+ public:
+  /**
+   * @brief SetInteriorVehicleDataRequest class constructor
+   *
+   * @param message Message from mobile
+   **/
+  SetInteriorVehicleDataRequest(const application_manager::MessagePtr& message,
+                                CANModuleInterface& can_module);
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AByzhynar please add description for `can_module` param in ctor

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * @brief Execute command
+   */
+  virtual void Execute();
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AByzhynar could be `FINAL` as it's already declared as virtual in base command

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                                             std::string>& event);
+
+  /**
+   * @brief SetInteriorVehicleDataRequest class destructor
+   */
+  virtual ~SetInteriorVehicleDataRequest();
+
+ protected:
+  virtual std::string ModuleType(const Json::Value& message);
+  virtual std::vector<std::string> ControlData(const Json::Value& message);
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AByzhynar `ModuleType` and `ControlData` also could be FINAL

File: src/components/application_manager/src/application_manager_impl.cc [hunk 6]
```diff
@@ -2074,9 +2074,54 @@ bool ApplicationManagerImpl::ConvertSOtoMessage(
...
+    case ProtocolVersion::kV2: {
+      const bool conversion_result =
+          formatters::CFormatterJsonSDLRPCv2::fromString(
+              message.json_message(),
+              so,
+              message.function_id(),
+              message.type(),
+              message.correlation_id());
+      if (!conversion_result) {
+        return INVALID_JSON;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/application_manager_impl.cc:2093]: @AByzhynar I think it would be good to add some logs for this function to see what was output result if you have few different scenarios in one function.

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "can_cooperation/commands/set_interior_vehicle_data_request.h"
+#include <algorithm>
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc:34]: @AByzhynar please move cpp headers before own headers

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+  LOG4CXX_DEBUG(logger_,
+                "HMI response validation result is " << validate_result);
+  const Json::Value value =
+      MessageHelper::StringToValue(hmi_response.json_message());
+
+  std::string result_code;
+  std::string info;
+
+  bool success = false;
+  if (IsMember(value, kError) && IsMember(value[kError], kCode) &&
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc]: @AByzhynar as I could see this part of logic is implemented in `BaseCommandRequest::ParseResultCode`
Why did you move it here?

**Review by AKalinich-Luxoft — CHANGES_REQUESTED** [review]

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc (hunk 7 above)
Comment (AByzhynar) [reply src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc:34]: @AKalinich-Luxoft This order is correct according to Coding Style used in project.

File: src/components/application_manager/include/application_manager/service.h (hunk 1 above)
Comment (LuxoftAKutsan) [reply src/components/application_manager/include/application_manager/service.h:230]: @AKalinich-Luxoft different use cases requires different behavior. Sometimes we need not only bool

File: src/components/application_manager/src/application_manager_impl.cc (hunk 6 above)
Comment (LuxoftAKutsan) [reply src/components/application_manager/src/application_manager_impl.cc:2093]: @AKalinich-Luxoft I think that logs here redundant we can check the branch of execution according return value

File: src/components/can_cooperation/src/commands/get_interior_vehicle_data_request.cc (hunk 2 above)
Comment (AByzhynar) [reply src/components/can_cooperation/src/commands/get_interior_vehicle_data_request.cc:81]: @AKalinich-Luxoft Yes. It's enough as this check is already included into Validation by schema.

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    module_type_and_data_match = !IsMember(module_data, kRadioControlData);
+  }
+
+  if (module_type_and_data_match) {
+    SendRequest(functional_modules::hmi_api::set_interior_vehicle_data,
+                request_params,
+                true);
+  } else {
+    LOG4CXX_WARN(logger_, "Request module type & data mismatch!");
+    SendResponse(false, result_codes::kInvalidData, "");
```
Comment (LuxoftAKutsan) [inline src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc]: @AByzhynar info parameter is not managed by requirements, but it would be useful to send information about cause of invalid data

File: src/components/application_manager/include/application_manager/service.h (hunk 1 above)
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/service.h:230]: @AKalinich-Luxoft I'ts just in this particular case. It was done for flexibility. This method used in other parts of code and any result code can be used.

**Review by LuxoftAKutsan — APPROVED** [review]

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc (hunk 9 above)
Comment (AByzhynar) [reply src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc]: @AKalinich-Luxoft Agree. Will send the same info to mobile.

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h (hunk 3 above)
Comment (AByzhynar) [reply src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AKalinich-Luxoft Done in c91395b773d3f52252cd8d1a2d5939daec2f4ea4

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h (hunk 4 above)
Comment (AByzhynar) [reply src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AKalinich-Luxoft Done in c91395b773d3f52252cd8d1a2d5939daec2f4ea4

File: src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h (hunk 5 above)
Comment (AByzhynar) [reply src/components/can_cooperation/include/can_cooperation/commands/set_interior_vehicle_data_request.h]: @AKalinich-Luxoft Done in c91395b773d3f52252cd8d1a2d5939daec2f4ea4

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc (hunk 8 above)
Comment (AByzhynar) [reply src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc]: @AKalinich-Luxoft Done in c91395b773d3f52252cd8d1a2d5939daec2f4ea4

File: src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc (hunk 9 above)
Comment (AByzhynar) [reply src/components/can_cooperation/src/commands/set_interior_vehicle_data_request.cc]: @LuxoftAKutsan  Done in c91395b773d3f52252cd8d1a2d5939daec2f4ea4

**Review by AKalinich-Luxoft — APPROVED** [review]

**Comment by AByzhynar** [comment]
@OlesiaV Please review part related to API changes

**Review by OlesiaV — APPROVED** [review]

**Review by LitvinenkoIra — APPROVED** [review]
